freeipa/src/main/java/com/sequenceiq/freeipa/converter/stack/StackToCreateFreeIpaRequestConverter.java

```diff
@@ -0,0 +1,393 @@
+        request.setBackup(getBackup(source));
+        request.setTags(getTags(source.getTags()));
+        request.setUseCcm(source.getUseCcm());
+        request.setTunnel(source.getTunnel());
+        request.setVariant(source.getPlatformvariant());
+
+        return request;
+    }
+
+    private String getName(Stack stack) {
+        String name = stack.getName();
+        if (stack.isDeleteCompleted()) {
+            int lastDelimeterIndex = name.lastIndexOf(DELETED_NAME_DELIMETER);
+            if (lastDelimeterIndex != -1) {
+                name = name.substring(0, lastDelimeterIndex);
+            }
+        }
+        return name;
+    }
+
+    private PlacementRequest getPlacementRequest(Stack stack) {
+        PlacementRequest placementRequest = new PlacementRequest();
+        placementRequest.setRegion(stack.getRegion());
+        placementRequest.setAvailabilityZone(stack.getAvailabilityZone());
+        return placementRequest;
+    }
+
+    private InstanceGroupRequest getInstanceGroupRequst(InstanceGroup instanceGroup) {
+        InstanceGroupRequest request = new InstanceGroupRequest();
+        request.setInstanceTemplateRequest(getInstanceTemplateRequest(instanceGroup.getTemplate()));
+        request.setName(instanceGroup.getGroupName());
+        request.setNetwork(getInststanceGroupNetworkRequest(instanceGroup.getInstanceGroupNetwork()));
+        request.setNodeCount(instanceGroup.getNodeCount());
+        request.setSecurityGroup(getSecurityGroupRequest(instanceGroup.getSecurityGroup()));
+        request.setType(instanceGroup.getInstanceGroupType());
+        return request;
+    }
+
+    private InstanceTemplateRequest getInstanceTemplateRequest(Template template) {
+        InstanceTemplateRequest request = null;
+        if (template != null) {
+            request = new InstanceTemplateRequest();
+            request.setInstanceType(template.getInstanceType());
+            Integer spotPercentage = template.getAttributes().getValue(AwsInstanceTemplate.EC2_SPOT_PERCENTAGE);
+            Double maxPrice = template.getAttributes().getValue(AwsInstanceTemplate.EC2_SPOT_MAX_PRICE);
+            if (spotPercentage != null) {
+                AwsInstanceTemplateParameters aws = new AwsInstanceTemplateParameters();
+                AwsInstanceTemplateSpotParameters spot = new AwsInstanceTemplateSpotParameters();
+                spot.setPercentage(spotPercentage);
+                if (maxPrice != null) {
+                    spot.setMaxPrice(maxPrice);
+                }
+                aws.setSpot(spot);
+                request.setAws(aws);
+            }
+            VolumeRequest volumeRequest = new VolumeRequest();
+            volumeRequest.setType(template.getVolumeType());
+            volumeRequest.setCount(template.getVolumeCount());
+            volumeRequest.setSize(template.getVolumeSize());
+            request.setAttachedVolumes(Optional.of(volumeRequest)
```

Comment:
Why are we using `Optional.of` here? Is it just to have access to the `filter` method?

Reply:
`Optional.of` allows the `filter()` to be applied.